test/shared/src/main/scala/zio/test/TestAspect.scala

```diff
@@ -193,13 +204,16 @@ object TestAspect extends TimeoutVariants {
   /**
    * An aspect that retries a test until success, without limit.
    */
-  val eventually: TestAspectPoly =
-    new TestAspect.PerTest[Nothing, Any, Nothing, Any, Nothing, Any] {
+  val eventually
+    : TestAspect[Nothing, TestClock with TestConsole with TestRandom with TestSystem, Nothing, Any, Nothing, Any] = {
```

Comment:
I am wondering if we just want to constrain it to `TestEnvironment` here and in the other combinators. It is adding a little bit of an additional constraint but it would make these type signatures less fiercesome and I don't really see these combinators being used with having implementations of all the test services and not the test environment. Also, conceptually we are trying to reset the entire test environment (e.g. if there was ever anything else added to the test environment that maintained state we would reset that too).

Reply:
Would it make sense to introduce something like:

```scala
type TestAspectTE = TestAspect[Nothing, TestClock with TestConsole with TestRandom with TestSystem, Nothing, Any, Nothing, Any]
```

to describe "at least" a test environment?

Reply:
I just changed it to use
```scala
type ResTestEnv = TestClock with TestConsole with TestRandom with TestSystem
def restoreTestEnvironment: TestAspect[Nothing, ResTestEnv, Nothing, Any, Nothing, Any] = ???
```
but I think your solution is cleaner

Reply:
@jdegoes  Yes, we could definitely do that. We could also do something like:

```scala
type ZTestEnv = TestClock with TestConsole with TestRandom with TestSystem
```

Reply:
Ah, I like that... probably it's already repeated in several places.

Maybe even make `TestAspectAtLeastR` type alias so we can plug in `ZTestEnv` and other things in other cases.

Reply:
Yes that is a good idea. We don't need to do it right now but we could also think of some other aliases like that. `TestAspectPoly` is really nice right now but if the test aspect isn't fully polymorphic you go straight from that to six type parameters.

Reply:
![image](https://user-images.githubusercontent.com/30517573/70357568-54e5de80-18cb-11ea-8319-019878dc0dc3.png)

```scala
  type ZTestEnv = TestClock with TestConsole with TestRandom with TestSystem
  type TestAspectTE =
    TestAspect[Nothing, ZTestEnv, Nothing, Any, Nothing, Any]
```

then we get
```scala
def retry[R0 <: ZTestEnv, E0, S0](schedule: Schedule[R0, TestFailure[E0], S0]) = ???
def restoreTestEnvironment: TestAspectTE = ???
```